stores/sqlstore.go

```diff
@@ -0,0 +1,1743 @@
+			err = sqlStmtError(sqlDeleteClient, err)
+			break
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &Client{spb.ClientInfo{ID: clientID, HbInbox: hbInbox}}, nil
+}
+
+// DeleteClient implements the Store interface
+func (s *SQLStore) DeleteClient(clientID string) error {
+	s.Lock()
+	_, err := s.preparedStmts[sqlDeleteClient].Exec(clientID)
+	if err != nil {
+		err = sqlStmtError(sqlDeleteClient, err)
+	}
+	s.Unlock()
+	return err
+}
+
+// timeTick updates the store's time in nanosecond at regular
+// interval. The time is used in Lookup() to compensate for possible
+// delay in expiring messages. The Lookup() will check the message's
+// expiration time against the time captured here. If it is expired
+// even though it is still in the database, Lookup() will return nil.
+func (s *SQLStore) timeTick() {
+	defer s.wg.Done()
+	timer := time.NewTicker(sqlTimeTickInterval)
+	for {
+		select {
+		case <-s.doneCh:
+			timer.Stop()
+			return
+		case <-timer.C:
+			atomic.StoreInt64(&s.nowInNano, time.Now().UnixNano())
+		}
+	}
+}
+
+// Close implements the Store interface
+func (s *SQLStore) Close() error {
+	s.Lock()
+	if s.closed {
+		s.Unlock()
+		return nil
+	}
+	s.closed = true
+	// This will cause MsgStore's and SubStore's to be closed.
+	err := s.close()
+	db := s.db
+	wg := &s.wg
+	// Signal background go-routines to quit
+	if s.doneCh != nil {
+		s.doneCh <- struct{}{}
+	}
+	if s.ssFlusher != nil {
+		s.ssFlusher.doneCh <- struct{}{}
+	}
+	s.Unlock()
+
+	// Wait for go routine(s) to finish
+	wg.Wait()
+
+	s.Lock()
+	for _, ps := range s.preparedStmts {
+		if lerr := ps.Close(); lerr != nil && err == nil {
+			err = lerr
+		}
+	}
+	if db != nil {
+		if lerr := db.Close(); lerr != nil && err == nil {
+			err = lerr
+		}
+	}
+	s.Unlock()
+	return err
+}
+
+////////////////////////////////////////////////////////////////////////////
+// SQLMsgStore methods
+////////////////////////////////////////////////////////////////////////////
+
+func (mc *sqlMsgsCache) add(msg *pb.MsgProto, data []byte) {
+	cachedMsg := mc.free
+	if cachedMsg != nil {
+		mc.free = cachedMsg.next
+		cachedMsg.next = nil
+		// Remove old message from the map
+		delete(mc.msgs, cachedMsg.msg.Sequence)
+	} else {
+		cachedMsg = &sqlCachedMsg{}
+	}
+	cachedMsg.msg = msg
+	cachedMsg.data = data
+	mc.msgs[msg.Sequence] = cachedMsg
+	if mc.head == nil {
+		mc.head = cachedMsg
+	} else {
+		mc.tail.next = cachedMsg
+	}
+	mc.tail = cachedMsg
+}
+
+func (mc *sqlMsgsCache) transferToFreeList() {
+	if mc.tail != nil {
+		mc.tail.next = mc.free
+		mc.free = mc.head
+	}
+	mc.head = nil
+	mc.tail = nil
+}
+
+func (mc *sqlMsgsCache) pop() *sqlCachedMsg {
+	cm := mc.head
+	if cm != nil {
+		delete(mc.msgs, cm.msg.Sequence)
+		mc.head = cm.next
+		if mc.head == nil {
+			mc.tail = nil
+		}
+	}
+	return cm
+}
+
+// Store implements the MsgStore interface
+func (ms *SQLMsgStore) Store(data []byte) (uint64, error) {
+	ms.Lock()
+	defer ms.Unlock()
+	seq := ms.last + 1
```

Comment:
It is an `uint64`. If it were to overflow, the software would be unusable anyway, user would have to remove files and start from scratch. Note that with `uint64` it will take a lonnnnnnng time before that happens.